File: iosApp/iosApp/LoginScreen.swift
//
//  LoginScreen.swift
//  iosApp
//
//  Created by Felix Fischer on 23.03.25.
//  Copyright © 2025 orgName. All rights reserved.
//

import SwiftUI
import ComposeApp

struct LoginScreen: View {
    let viewModel: LoginViewModel
    let localizables = LoginScreenLocalizables()
    @FocusState private var focusedField: Field?
    
    var body: some View {
        StatefulView(
            viewModel: viewModel,
            buildAlert: { (error: LoginScreenError) in
                error.asBasicAlert()
            },
            content: content(data:)
        )
        .navigationTitle(localizables.navigationTitle.localized)
        .background {
            DesignSystem.background
                .ignoresSafeArea()
        }
    }
    
    private func content(data: LoginViewModel.Data) -> AnyView {
        VStack {
            Spacer()
            
            VStack(spacing: 56) {
                CustomTextField(
                    label: localizables.usernameLabel.localized,
                    text: data.username,
                    onChange: viewModel.updateUsername(username:)
                )
                .textFieldStyle(DesignSystem.CustomTextFieldStyle())
                .focused($focusedField, equals: .username)
                .submitLabel(.next)
                .onSubmit {
                    focusedField = .password
                }

                CustomSecureField(
                    label: localizables.passwordLabel.localized,
                    text: data.password,
                    onChange: viewModel.updatePassword(password:)
                )
                .textFieldStyle(DesignSystem.CustomTextFieldStyle())
                .focused($focusedField, equals: .password)
                .submitLabel(.go)
                .onSubmit(viewModel.login)

                Button(action: viewModel.register) {
                    Text(localizables.registerHintText.localized)
                        .font(.footnote)
                }
            }

            Spacer()

            DesignSystem.PrimaryButton(
                title: localizables.loginButtonTitle.localized,
                action: viewModel.login
            )
        }
        .padding(DesignSystem.Padding.large)
        .eraseToAnyView()
    }
    
    private enum Field {
        case username
        case password
    }
}

fileprivate typealias LoginScreenError = ViewModelStateCombinedError<LoginViewModel.Error, NetworkError>

File: composeApp/src/commonMain/kotlin/felix/livinglink/ui/login/LoginScreen.kt
package felix.livinglink.ui.login

import LoginScreenLocalizables
import androidx.compose.foundation.layout.imePadding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import felix.livinglink.ui.common.BackAwareScaffold
import felix.livinglink.ui.common.state.StatefulView

@Composable
fun LoginScreen(viewModel: LoginViewModel) {
    BackAwareScaffold(
        navigator = viewModel.navigator,
        title = LoginScreenLocalizables.navigationTitle()
    ) { innerPadding ->
        StatefulView(
            viewModel = viewModel,
            modifier = innerPadding
        ) { data ->
            LoginScreenContent(
                data = data,
                viewModel = viewModel,
                modifier = Modifier.imePadding()
            )
        }
    }
}